### ISSUE TITLE
Support the 'HttpOnly' flag when setting cookies

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5911]
|<img alt="reed's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| reed reported|
|-|-|
|Trac ID|trac#5911|
|Type|enhancement|
|Created|2012-08-24 02:52:15Z|
|Branch|https://github.com/twisted/twisted/tree/web-http-cookie-httponly-5911-5|

The Twisted 'web' module doesn't support the addition of the 'HttpOnly' flag when setting a cookie.

http://www.owasp.org/index.php/HTTPOnly

Attachments:

* [support_httponly_flag.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/69e7d37ab0e97f2f3821d72e32edc0225351207d.diff) (1332 bytes) - added by reed on 2012-08-24 02:53:45Z - Add support for HttpOnly flag in cookies
* [support_httponly_flag.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/82530c541082f544a3e185f50e6a2baf103399df.diff) (2636 bytes) - added by reed on 2012-08-24 03:14:00Z - Same as above but includes a test
* [5911-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/5c1787cda1fbb30df76a0286a3fa84d083251976.patch) (4230 bytes) - added by d.vinella on 2015-10-21 08:23:15Z - 
* [5911-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/858c4edcadcee82d024c0494b9ac353a0f55981e.patch) (4382 bytes) - added by d.vinella on 2015-10-21 20:38:24Z - 
* [5911-5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/c21c0b5c91b7ed4736197432175744d573e3d036.patch) (4439 bytes) - added by d.vinella on 2015-10-22 08:49:05Z - 
* [5911-6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/4971209e6e9a1193e30271e29cfedb760cc7c7b4.patch) (4404 bytes) - added by d.vinella on 2015-10-24 11:26:05Z - 
* [5911-7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87ec635380174d9a74c59851534f4fbe49bd885f/17a87466b9072a4b387a894f0551e70db476fa40.patch) (4362 bytes) - added by d.vinella on 2015-10-27 08:45:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5911 5911
type__enhancement enhancement
reporter__reed reed
priority__normal normal
milestone__ 
branch__branches_web_http_cookie_httponly_5911_5 branches/web-http-cookie-httponly-5911-5
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1345776735000000 1345776735000000
changetime__1446287447383172 1446287447383172
version__None None
owner__hawkowl hawkowl
cc__jknight cc__d.vinella@...
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="reed's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|reed commented|
|-|-|

While Python has supported the 'HttpOnly' flag in cookies for quite a while (http://bugs.python.org/issue1638033), it doesn't look like there's a 'httponly' attribute, at least according to http://docs.python.org/library/cookielib.html#cookie-objects. I will follow-up with that to see if that's true, and if so, get it fixed upstream.

New patch with a test coming up.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to reed|
|-|-|

Thanks for your patch! There is one main problem:

1. Your test doesn't match your change. The test applies to the client, not server Request.addCookie.

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|d.vinella removed owner|
|-|-|

A new patch with correct tests. 
Please review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

(In [45878]) Branching to web-http-cookie-httponly-5911.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to d.vinella|
|-|-|

Many thanks for you work on this!

------

Can you please also document the types of the parameters ?

We want them to be bytes... so we should let users know that no explicit encoding is done at this level and that they should not pass text/Unicode.

Also, the test should be written with explict bytes marker b'key' instead of just 'b'

We should try to write code so that it is Py3 friendly ... and in Py3 all texts are now Unicode.

---------

Maybe rename  test_addCookie - > test_addCookieWithMinimumArguments

And have a docstring like: Add a 'Set-Cookie' header with just key and value.

--------

Similar: test_addCookieWithAttibutes - > test_addCookieWithAllArguments

Docstring: Add a 'Set-Cookie' header with key and value and all the supported options.

--------

assertIn is pretty weak... can we use assertEqual ?
I case the order is not the same, can we use a cookie parser to read the result and check it in that way.

My problem is that assertIn tests will still pass even there is junk between the asserted bits. ...
for example 'foo=bar ; junk HttpOnly' 


-------

Please check my comments and resubmit.

I have sent your previous patch to the builders.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|d.vinella removed owner|
|-|-|

Thanks for your review, Adi. 
Here it is a new patch with all the changes. Please review

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

(In [45886]) Branching to web-http-cookie-httponly-5911-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to d.vinella|
|-|-|

Changes look good but test are failing with the latest patch.

Please check the buildbot results https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/web-http-cookie-httponly-5911-2

Please fix the test and resubmit.

Since the http module is ported on Py3 we need to make sure changes work on both python 2.7 and a recent py3 version... ex 3.4

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|d.vinella removed owner|
|-|-|

This patch passes tests on py3.5 too, please review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

(In [45890]) Branching to web-http-cookie-httponly-5911-3.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set owner to d.vinella|
|-|-|

There are still some errors on py 3.4

```
builtins.TypeError: unsupported operand type(s) for %: 'bytes' and 'tuple'

twisted.web.test.test_http.RequestTests.test_addCookieWithMinimumArguments
```

Please try to fix them and submit a new patch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|d.vinella removed owner|
|-|-|

Here it is a new patch, please review!

One concern I'd like to raise is the potential api breaking change for the
max_age param: previously it would accept strings and ints, it was undocumented
and never used anywhere in the twised (nor nevow) codebase. Now it is more
restrictive and it accepts ints only. If anyone out there is using max_age
with a string (or bytes), it will break for her. I don't know if some 
isinstance trick is stylistically acceptable or should be avoided so any 
advice is welcome for this review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

(In [45949]) Branching to web-http-cookie-httponly-5911-4.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban commented|
|-|-|

Thanks for your patch. It looks good.

I don't know how to proceed with the max_age.

I would say that we should continue to accept str/bytes but raise a deprecation warning.

Also, maybe we can deprecate the whole max_age argument and add a new maxAge argument to comply with the coding standard.

I will open a discussion on the mailing list to get some feedback.

I am leaving this as review so that other can provide their feedback.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl set owner to d.vinella|
|-|-|

Hi! Thanks for the patch.

I think we can sidestep this issue because at https://github.com/twisted/twisted/blob/trunk/twisted/web/http.py#L1010 , it is changed from a str to bytes as the cookies are added. This ensures that all keys/values are ASCII-encodable as well, while concatenating bytes has no such check -- and I believe things get a bit wacky with non-ASCII headers, so we should keep the check anyway.

Please revert the str -> bytes changes, and document the API as taking str for now, and we 
can deal with addCookie taking str and not bytes in another ticket (I filed [#8067](https://github.com/twisted/twisted/issues/8067) for it). 

The rest of the API docs and changes look fine, just put up another patch with the str/bytes changes reverted as above, and resubmit for review.

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|d.vinella removed owner|
|-|-|

Thank you for your review, hawkowl. Here it is the patch with str inputs. 
Please review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban commented|
|-|-|

(In [46001]) Branching to web-http-cookie-httponly-5911-5.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban commented|
|-|-|

Thanks for the patch.

I have sent the patch to buildbots and they look green.

I will let hawkowl to the final review... as at this stage I am confused regarding the handling of str/unicode and bytes in Twisted Web.

Thanks again for your contribution.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl set owner to d.vinella|
|-|-|

Hi, thanks for the updated patch!

The code is all good, but I'd like the topfile to mention why this is a good change -- what benefit httpOnly gives ```twisted.web``` users, so they know why they should use our spiffy new features :)

You don't have to upload another patch, just comment with it here and I'll amend it before merging. :)

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|d.vinella removed owner|
|-|-|

Thank you for the review, hawkowl,
the extendend topfile line could be like this

```
twisted.web.http.Request.addCookie now supports httpOnly attribute to set 
cookies that the browser will not expose through channels other than HTTP 
and HTTPS requests (i.e. they will not be accessible through JavaScript).
```

plus an additional, optional, extension

```
These are are suitable to securely store auth tokens.
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_23"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Thanks! I'll merge this when I get back into work tomorrow :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_24"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46010]) Merge web-http-cookie-httponly-5911-5: Support the 'HttpOnly' flag when setting cookies

Author: d.vinella
Reviewers: adiroiban, hawkowl
Fixes: [#5911](https://github.com/twisted/twisted/issues/5911)
